File: packages/flutter-im-ui/test/locate_orchestration_vectors_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flare_im_ui/flare_im_ui.dart';
import 'package:flutter_test/flutter_test.dart';

/// The shared locate table (`spec/locate-orchestration-vectors.json`) is the one trip four kits make when a
/// quote names a message that is not loaded. Each case scripts a run; the expectations count the whole trip,
/// so a kit that asks or pages a different number of times fails even when it lands on the same answer.
void main() {
  final table =
      jsonDecode(
            File(
              '../../spec/locate-orchestration-vectors.json',
            ).readAsStringSync(),
          )
          as Map<String, dynamic>;
  final cases = (table['cases'] as List).cast<Map<String, dynamic>>();

  test('the table is the budget this kit was built with', () {
    expect(table['maxPages'], flareLocateMaxPages);
    expect(table['settleAttempts'], flareLocateSettleAttempts);
    expect(cases.length, greaterThanOrEqualTo(13));
  });

  for (final vector in cases) {
    final id = vector['id'] as String;
    test('locate run: $id', () async {
      var pages = 0;
      var shows = 0;
      final historyPages = vector['historyPages'] as int;
      final foundAfterPages = vector['foundAfterPages'] as int?;
      final failAtPage = vector['failAtPage'] as int?;
      final cancelAfterShows = vector['cancelAfterShows'] as int?;
      final visibleFrom = vector['visibleFromShow'] as int? ?? 1;
      final outcome = await flareLocateMessage(
        isCurrent: () => cancelAfterShows == null || shows < cancelAfterShows,
        showInList: () {
          shows += 1;
          if (foundAfterPages == null) return false;
          return pages >= foundAfterPages && shows >= visibleFrom;
        },
        hasOlder: () => pages < historyPages,
        readOlder: () {
          pages += 1;
          return failAtPage == null || pages != failAtPage;
        },
        settle: () {},
      );
      final expected = vector['expected'] as Map<String, dynamic>;
      expect(outcome.name, expected['outcome'], reason: '$id outcome');
      expect(pages, expected['pagesRead'], reason: '$id pages read');
      expect(shows, expected['showCalls'], reason: '$id asks');
    });
  }
}

File: packages/android-im-ui/src/test/kotlin/com/flare/im/ui/LocateOrchestrationVectorsTest.kt
package com.flare.im.ui

import java.io.File
import kotlinx.coroutines.runBlocking
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * The shared locate table (`spec/locate-orchestration-vectors.json`) is the one trip four kits make when a
 * quote names a message that is not loaded. Each case scripts a run; the expectations count the whole trip,
 * so a kit that asks or pages a different number of times fails even when it lands on the same answer.
 */
class LocateOrchestrationVectorsTest {
    @Test fun theTableIsTheBudgetThisKitWasBuiltWith() {
        val table = FlareJson.parse(vectorsFile().readText()) as Map<*, *>
        assertEquals(FLARE_LOCATE_MAX_PAGES, (table["maxPages"] as Number).toInt())
        assertEquals(FLARE_LOCATE_SETTLE_ATTEMPTS, (table["settleAttempts"] as Number).toInt())
    }

    @Test fun everyRunEndsTheSameWayHere() = runBlocking {
        val table = FlareJson.parse(vectorsFile().readText()) as Map<*, *>
        val cases = table["cases"] as List<*>
        assertTrue(cases.size >= 13, "the shared table lost cases: ${cases.size}")
        for (raw in cases) {
            val case = raw as Map<*, *>
            val id = case["id"] as String
            val historyPages = (case["historyPages"] as Number).toInt()
            val foundAfterPages = (case["foundAfterPages"] as Number?)?.toInt()
            val failAtPage = (case["failAtPage"] as Number?)?.toInt()
            val cancelAfterShows = (case["cancelAfterShows"] as Number?)?.toInt()
            val visibleFrom = (case["visibleFromShow"] as Number?)?.toInt() ?: 1
            var pages = 0
            var shows = 0
            val outcome = flareLocateMessage(
                showInList = {
                    shows += 1
                    foundAfterPages != null && pages >= foundAfterPages && shows >= visibleFrom
                },
                hasOlder = { pages < historyPages },
                readOlder = {
                    pages += 1
                    failAtPage == null || pages != failAtPage
                },
                settle = {},
                isCurrent = { cancelAfterShows == null || shows < cancelAfterShows },
            )
            val expected = case["expected"] as Map<*, *>
            assertEquals(expected["outcome"], outcome.name.replaceFirstChar { it.lowercase() }, "$id outcome")
            assertEquals((expected["pagesRead"] as Number).toInt(), pages, "$id pages read")
            assertEquals((expected["showCalls"] as Number).toInt(), shows, "$id asks")
        }
    }
}

/** The table lives beside the four kits, so find the repository root from the module's directory. */
private fun vectorsFile(): File {
    var dir: File? = File(".").absoluteFile
    while (dir != null) {
        val candidate = File(dir, "spec/locate-orchestration-vectors.json")
        if (candidate.isFile) return candidate
        dir = dir.parentFile
    }
    error("spec/locate-orchestration-vectors.json not found above ${File(".").absolutePath}")
}
